cEP0002.md

```diff
@@ -0,0 +1,166 @@
+very well as it sometimes has synchronization problems. Also if local bears
+fail to run it stalls the global bears. Also we have a permanent logger thread.
+An extra logger thread consumes time and resources of the system which we can
+avoid.
+
+The new proposal aims to eliminate the use of the queuing control signals.
+Utilising a multiprocessing pool to generate the results from the bears. No
+longer will the local bears be dependent on global bears to finish. The logging
+will also be handled by the main thread itself instead of utilising an extra
+thread.
+
+# Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. coala will have a core multiprocessing pool.
+2. The `Bear` class will get two new methods. `__execute_task()` and
+`generate_tasks()`.
+    * The `generate_tasks()` is resposible for generating the correct
+    arguments for the upcoming tasks.
+    * The `__execute_task()` method is called in parallel by the
+    multiprocessing pool on as many available threads.
+3. Each bear will load the core coala pool with different types of tasks.
+    * Global bears fire tasks that process all files.
+    * Local bears fire tasks for each individual file.
+6. As the task runs it places it's generated results into a result queue
+during runtime and continues to do so till the task completes.
+7. The main coala process waits on the result queue to get a result and then
+proceeds to process it and log it when it arrives.
+
+
+# A Prototype
+
+A rough look at how the new local bear class implementation will look:
+
+```python
+class LocalBear:
+    def __init__(self, section, result_queue, file_dict):
+        self.section = section
+        self.result_queue = result_queue
+        self.file_dict = file_dict
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def __execute_task(self, task):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        for result in self.run(task, **self.kwargs):
+            self.result_queue.put(result)
+
+    def run(self,
+            file_proxy
+            *args,
+            dependency_results=None,
+            **kwargs):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        raise NotImplementedError("This function has to be implemented for a "
+                                  "runnable bear.")
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the files.
+        """
+        for fileproxy in self.file_dict.values():
+            yield fileproxy
+```
+
+A test bear:
+
+```python
+class TestBear(LocalBear):
+    def run(self, file_proxy):
```

Comment:
when a bear would use the old signature, then our file-proxy objects would get meaningless, as we need to invoke the functions from the file-proxy object to load the parameters with content.
```
def run(self, file_proxy):
    str(file_proxy) # <-- file contents as a string, now loaded from file when run first

def run(self, filename, file):
    # for compatability, file would need the presplitted lines:
    if does_not_meet_convention(filename):
        yield Result(...)
    # but we don't use them at all, the file-proxy object under the hood would load
    # / split the lines anyway, and that everytime for every file, rendering this object
    # totally meaningless^^
```